src/core/server/saved_objects/service/lib/legacy_url_aliases/delete_legacy_url_aliases.ts

```diff
@@ -0,0 +1,112 @@
+   */
+  deleteBehavior: 'inclusive' | 'exclusive';
+}
+
+/**
+ * Deletes legacy URL aliases that point to a given object.
+ *
+ * Note that aliases are only created when an object is converted to become share-capable, and each targetId is deterministically generated
+ * with uuidv5 -- this means that the chances of there actually being _multiple_ legacy URL aliases that target a given type/ID are slim to
+ * none. However, we don't always know exactly what space an alias could be in (if an object exists in multiple spaces, or in all spaces),
+ * so the most straightforward way for us to ensure that aliases are reliably deleted is to use updateByQuery, which is what this function
+ * does.
+ *
+ * @internal
+ */
+export async function deleteLegacyUrlAliases(params: DeleteLegacyUrlAliasesParams) {
+  const { mappings, registry, client, getIndexForType, type, id, namespaces, deleteBehavior } =
+    params;
+
+  if (namespaces.includes(ALL_NAMESPACES_STRING)) {
+    throwError(type, id, '"namespaces" cannot include the * string');
+  }
+
+  // Legacy URL aliases cannot exist in the default space; filter that out
+  const filteredNamespaces = namespaces.filter(
+    (namespace) => namespace !== DEFAULT_NAMESPACE_STRING
+  );
+  if (!filteredNamespaces.length && deleteBehavior === 'inclusive') {
+    // nothing to do, return early
+    return;
+  }
+
+  const { buildNode } = esKuery.nodeTypes.function;
+  const match1 = buildNode('is', `${LEGACY_URL_ALIAS_TYPE}.targetType`, type);
+  const match2 = buildNode('is', `${LEGACY_URL_ALIAS_TYPE}.targetId`, id);
+  const kueryNode = buildNode('and', [match1, match2]);
+
+  try {
+    await client.updateByQuery(
+      {
+        index: getIndexForType(LEGACY_URL_ALIAS_TYPE),
+        refresh: false, // This could be called many times in succession, intentionally do not wait for a refresh
+        body: {
+          script: {
+            // Intentionally use one script source with variable params to take advantage of ES script caching
+            source: `
+              if (params['namespaces'].indexOf(ctx._source['${LEGACY_URL_ALIAS_TYPE}']['targetNamespace']) > -1) {
+                ctx.op = params['matchTargetNamespaceOp'];
+              } else {
+                ctx.op = params['notMatchTargetNamespaceOp'];
+              }
+            `,
+            params: {
+              namespaces: filteredNamespaces,
+              matchTargetNamespaceOp: deleteBehavior === 'inclusive' ? 'delete' : 'noop',
+              notMatchTargetNamespaceOp: deleteBehavior === 'inclusive' ? 'noop' : 'delete',
+            },
+            lang: 'painless',
+          },
+          conflicts: 'proceed',
+          ...getSearchDsl(mappings, registry, {
+            type: LEGACY_URL_ALIAS_TYPE,
+            kueryNode,
+          }),
+        },
+      },
+      { ignore: [404] }
+    );
+  } catch (err) {
+    const { error } = err.body;
```

Comment:
We're handling expected ES errors here, but an unexpected error may not have a `body` property, or that `body` property may not have an `error` property. If we encounter an error during this error handling, then we'll lose the origial error context and have a difficult time tracking down exactly what happened

Reply:
Good call, I found this function that the core ES client module provides:

https://github.com/elastic/kibana/blob/2cd007ebfc64249a0d54d387e61d7a284832f151/src/core/server/elasticsearch/client/configure_client.ts#L83-L93

I'll reuse it here